{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pickle\n",
    "import sys\n",
    "import os\n",
    "import datetime\n",
    "\n",
    "import sklearn\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.cross_validation import StratifiedKFold\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.multiclass import OneVsRestClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from sklearn.preprocessing import label_binarize\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Import data from the processing of data in the journal_training_word2vec notebook\n",
    "#df = pickle.load(open('/home/sarahwie/Documents/pubmed-nlp-research/C_article_replication_and_translational_classification/df.p', 'rb'))\n",
    "df = pickle.load(open('/mnt/mypartition/pubmed_nlp_research/C_article_replication_and_translational_classification/pickled_objects/dfJournals_try2.p', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(96673, 5)\n",
      "['sentiment' 'title' 'abstract' 'qualifier_terms' 'descriptor_terms']\n",
      "['T0' 'T1/T2' 'T3/T4']\n"
     ]
    }
   ],
   "source": [
    "print df.shape\n",
    "print df.columns.values\n",
    "print df['sentiment'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#for the journals VM dataframe, change the label of the 'sentiment' column to 'label'\n",
    "df.columns = ['label', 'title', 'abstract', 'qualifier_terms', 'descriptor_terms']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>qualifier_terms</th>\n",
       "      <th>descriptor_terms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T0</td>\n",
       "      <td>Quantification of the relative contributions o...</td>\n",
       "      <td>Amyotrophic lateral sclerosis (ALS) and fronto...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T0</td>\n",
       "      <td>Ubiquitin Specific Protease 36 (USP36) control...</td>\n",
       "      <td>Ubiquitination of the TrkA neurotrophin recept...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T0</td>\n",
       "      <td>Functional Diversity of Cytotoxic tRNase/immun...</td>\n",
       "      <td>Contact-dependent growth inhibition (CDI) is a...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T0</td>\n",
       "      <td>Identification of a distinct substrate binding...</td>\n",
       "      <td>The type III secretion system (T3SS) effector ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T0</td>\n",
       "      <td>Role of chondroitin sulfate (CS) modification ...</td>\n",
       "      <td>Protein tyrosine phosphatase receptor type Z (...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T0</td>\n",
       "      <td>Chemically programmed bispecific antibodies in...</td>\n",
       "      <td>Chemically programmed bispecific antibodies (b...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T0</td>\n",
       "      <td>Non-mutagenic Suppression of Enterocyte Ferrop...</td>\n",
       "      <td>Iron transfer across the basolateral membrane ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T0</td>\n",
       "      <td>Structural plasticity of cholesteryl ester tra...</td>\n",
       "      <td>Cholesteryl ester transfer protein (CETP) medi...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T0</td>\n",
       "      <td>Identification of a Membrane-Bound Prepore Spe...</td>\n",
       "      <td>Pore-forming toxins (PFT) are cytolytic protei...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T0</td>\n",
       "      <td>EspR-dependent ESAT-6 secretion of Mycobacteri...</td>\n",
       "      <td>Attenuation of M. bovis BCG strain is related ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  label                                              title  \\\n",
       "0    T0  Quantification of the relative contributions o...   \n",
       "1    T0  Ubiquitin Specific Protease 36 (USP36) control...   \n",
       "2    T0  Functional Diversity of Cytotoxic tRNase/immun...   \n",
       "3    T0  Identification of a distinct substrate binding...   \n",
       "4    T0  Role of chondroitin sulfate (CS) modification ...   \n",
       "5    T0  Chemically programmed bispecific antibodies in...   \n",
       "6    T0  Non-mutagenic Suppression of Enterocyte Ferrop...   \n",
       "7    T0  Structural plasticity of cholesteryl ester tra...   \n",
       "8    T0  Identification of a Membrane-Bound Prepore Spe...   \n",
       "9    T0  EspR-dependent ESAT-6 secretion of Mycobacteri...   \n",
       "\n",
       "                                            abstract qualifier_terms  \\\n",
       "0  Amyotrophic lateral sclerosis (ALS) and fronto...                   \n",
       "1  Ubiquitination of the TrkA neurotrophin recept...                   \n",
       "2  Contact-dependent growth inhibition (CDI) is a...                   \n",
       "3  The type III secretion system (T3SS) effector ...                   \n",
       "4  Protein tyrosine phosphatase receptor type Z (...                   \n",
       "5  Chemically programmed bispecific antibodies (b...                   \n",
       "6  Iron transfer across the basolateral membrane ...                   \n",
       "7  Cholesteryl ester transfer protein (CETP) medi...                   \n",
       "8  Pore-forming toxins (PFT) are cytolytic protei...                   \n",
       "9  Attenuation of M. bovis BCG strain is related ...                   \n",
       "\n",
       "  descriptor_terms  \n",
       "0                   \n",
       "1                   \n",
       "2                   \n",
       "3                   \n",
       "4                   \n",
       "5                   \n",
       "6                   \n",
       "7                   \n",
       "8                   \n",
       "9                   "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('ROUND', 1)\n",
      "Cleaning and parsing the training set abstracts...\n",
      "\n",
      "77337\n",
      "Creating the bag of words...\n",
      "\n",
      "5000\n",
      "Cleaning and parsing the test set abstracts...\n",
      "\n",
      "Adding tfidf weights and converting to Bag of Words...\n",
      "Training the random forest...\n",
      "3\n",
      "3\n",
      "Using the random forest to make sentiment label predictions...\n",
      "Scoring the test set\n",
      "{0: 0.97434903270299356, 1: 0.94420222164849055, 2: 0.91153291222011479}\n",
      "0:12:13\n"
     ]
    }
   ],
   "source": [
    "a = datetime.datetime.now().replace(microsecond=0)\n",
    "\n",
    "y = df['label'].values\n",
    "skf = StratifiedKFold(y, n_folds=5, shuffle=True)\n",
    "\n",
    "#include mesh terms in bag of words?\n",
    "mesh_terms = True\n",
    "#remove stopwords?\n",
    "stops = False\n",
    "#if we want to set max features or not for the BOWs\n",
    "maxF = 5000\n",
    "\n",
    "j = 1\n",
    "avgT0 = []\n",
    "avgT1T2 = []\n",
    "avgT3T4 = []\n",
    "for train_index, test_index in skf:\n",
    "    #get first test/train split and use\n",
    "    if j == 1:\n",
    "\n",
    "        print(\"ROUND\", j)\n",
    "        j = j + 1\n",
    "        #use the indexes to subset the df pandas dataframe\n",
    "        train1, test1 = df.iloc[train_index], df.iloc[test_index]\n",
    "\n",
    "        print \"Cleaning and parsing the training set abstracts...\\n\"\n",
    "        # Initialize an empty list to hold the clean reviews\n",
    "        clean_train_reviews = []\n",
    "\n",
    "        # Loop over each review; create an index i that goes from 0 to the length\n",
    "        # of the pandas df column\n",
    "        for i in range(train1[\"abstract\"].size):\n",
    "            # Call our function for each one, and add the result to the list of\n",
    "            # clean reviews\n",
    "            words = KaggleWord2VecUtility.review_to_words( train1[\"abstract\"].iloc[i], remove_stopwords=stops)\n",
    "            words = words + ' ' + KaggleWord2VecUtility.review_to_words( train1[\"title\"].iloc[i], remove_stopwords=stops)\n",
    "            if mesh_terms:\n",
    "                words = words + ' ' + KaggleWord2VecUtility.review_to_words( train1[\"descriptor_terms\"].iloc[i], remove_stopwords=stops)\n",
    "                words = words + ' ' + KaggleWord2VecUtility.review_to_words( train1[\"qualifier_terms\"].iloc[i], remove_stopwords=stops)\n",
    "            clean_train_reviews.append(words)\n",
    "\n",
    "        print len(clean_train_reviews)\n",
    "\n",
    "        #with tfidf word weighting\n",
    "        #uses L2 norm by default\n",
    "        print \"Creating the bag of words...\\n\"\n",
    "\n",
    "        # Initialize the \"TfidfVectorizer\" object, which is scikit-learn's\n",
    "        # bag of words tool.  \n",
    "        vectorizer = TfidfVectorizer(analyzer = \"word\",   \\\n",
    "                                 tokenizer = None,    \\\n",
    "                                 preprocessor = None, \\\n",
    "                                #already removed stopwords when preprocessing reviews into words\n",
    "                                 stop_words = None,\n",
    "                                max_features = maxF)\n",
    "\n",
    "        # fit_transform() does two functions: First, it fits the model\n",
    "        # and learns the vocabulary; second, it transforms our training data\n",
    "        # into feature vectors. The input to fit_transform should be a list of \n",
    "        # strings.\n",
    "        train_data_features = vectorizer.fit_transform(clean_train_reviews)\n",
    "        print len(vectorizer.get_feature_names())\n",
    "\n",
    "        # Numpy arrays are easy to work with, so convert the result to an \n",
    "        # array\n",
    "        train_data_features = train_data_features.toarray()\n",
    "\n",
    "        #Moving on to test set...\n",
    "        # Create an empty list and append the clean reviews one by one\n",
    "        clean_test_reviews = [] \n",
    "\n",
    "        print \"Cleaning and parsing the test set abstracts...\\n\"\n",
    "        for i in range(test1[\"abstract\"].size):    \n",
    "            words = KaggleWord2VecUtility.review_to_words( test1[\"abstract\"].iloc[i], remove_stopwords=stops)\n",
    "            words = words + ' ' + KaggleWord2VecUtility.review_to_words( test1[\"title\"].iloc[i], remove_stopwords=stops)\n",
    "            if mesh_terms:\n",
    "                words = words + ' ' + KaggleWord2VecUtility.review_to_words( test1[\"descriptor_terms\"].iloc[i], remove_stopwords=stops)\n",
    "                words = words + ' ' + KaggleWord2VecUtility.review_to_words( test1[\"qualifier_terms\"].iloc[i], remove_stopwords=stops)\n",
    "            clean_test_reviews.append(words)\n",
    "\n",
    "        print \"Adding tfidf weights and converting to Bag of Words...\"\n",
    "        # Get a bag of words for the test set, and convert to a numpy array\n",
    "        test_data_features = vectorizer.transform(clean_test_reviews)\n",
    "        test_data_features = test_data_features.toarray()\n",
    "\n",
    "        #------------------------------------------------------------------------------------------------------------    \n",
    "        print \"Training the random forest...\"\n",
    "\n",
    "        #binarize labels\n",
    "        x = label_binarize(train1['label'], classes=['T0', 'T1/T2', 'T3/T4'])\n",
    "        n_classes = x.shape[1]\n",
    "        print n_classes\n",
    "\n",
    "        x_test = label_binarize(test1['label'], classes=['T0', 'T1/T2', 'T3/T4'])\n",
    "\n",
    "        # Initialize a Random Forest classifier with 100 trees\n",
    "        forest = OneVsRestClassifier(RandomForestClassifier(n_estimators = 100, max_depth=1000))\n",
    "\n",
    "        # Fit the forest to the training set, using the bag of words as \n",
    "        # features and the sentiment labels as the response variable\n",
    "        #\n",
    "        # This may take a few minutes to run\n",
    "        forest = forest.fit( train_data_features, x )\n",
    "\n",
    "        #get the feature importance for each of the three fits\n",
    "        print len(forest.estimators_)\n",
    "        feats = []\n",
    "        for i in range(len(forest.estimators_)):\n",
    "            feats.append(forest.estimators_[i].feature_importances_)\n",
    "\n",
    "        print \"Using the random forest to make sentiment label predictions...\"\n",
    "        result = forest.predict(test_data_features)\n",
    "\n",
    "        print \"Scoring the test set\"\n",
    "\n",
    "        # Compute ROC curve and ROC area for each class\n",
    "        fpr = dict()\n",
    "        tpr = dict()\n",
    "        roc_auc = dict()\n",
    "        for i in range(n_classes):\n",
    "            fpr[i], tpr[i], _ = roc_curve(x_test[:, i], result[:, i])\n",
    "            roc_auc[i] = auc(fpr[i], tpr[i])\n",
    "\n",
    "        print roc_auc\n",
    "\n",
    "#     avgT0.append(roc_auc[0])\n",
    "#     avgT1T2.append(roc_auc[1])\n",
    "#     avgT3T4.append(roc_auc[2])\n",
    "\n",
    "# print sum(avgT0) / float(len(avgT0))\n",
    "# print sum(avgT1T2) / float(len(avgT1T2))\n",
    "# print sum(avgT3T4) / float(len(avgT3T4))\n",
    "\n",
    "b = datetime.datetime.now().replace(microsecond=0)\n",
    "print(b-a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### External validation on paper's dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_val = pickle.load(open('/mnt/mypartition/pubmed_nlp_research/C_article_replication_and_translational_classification/df.p', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(354, 6)\n",
      "['pmid' 'title' 'abstract' 'qualifier_terms' 'descriptor_terms' 'label']\n",
      "['T0' 'T1/T2' 'T3/T4' 'TX']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>qualifier_terms</th>\n",
       "      <th>descriptor_terms</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>23195993</td>\n",
       "      <td>Gene expression profiles in peripheral blood m...</td>\n",
       "      <td>Occupational exposure to nickel (Ni) is associ...</td>\n",
       "      <td>genetics metabolism epidemiology drug effects ...</td>\n",
       "      <td>Adult Asian Continental Ancestry Group Biomark...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25077433</td>\n",
       "      <td>SOX2 is a cancer-specific regulator of tumour ...</td>\n",
       "      <td>Although the principles that balance stem cell...</td>\n",
       "      <td>genetics metabolism pathology metabolism patho...</td>\n",
       "      <td>Animals Carcinoma, Squamous Cell Cell Line, Tu...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24107601</td>\n",
       "      <td>Imaging and cerebrospinal fluid biomarkers in ...</td>\n",
       "      <td>The pathophysiological process of Alzheimer's ...</td>\n",
       "      <td>cerebrospinal fluid genetics radionuclide imag...</td>\n",
       "      <td>Aged Aged, 80 and over Alzheimer Disease Amylo...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>24891029</td>\n",
       "      <td>Preliminary evidence of cognitive and brain ab...</td>\n",
       "      <td>To ascertain whether pediatric obesity without...</td>\n",
       "      <td>pathology physiopathology physiopathology</td>\n",
       "      <td>Adolescent Attention Brain Cognition Executive...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21691448</td>\n",
       "      <td>Obese Adolescents with Type 2 Diabetes Mellitu...</td>\n",
       "      <td>The rates of type 2 diabetes (T2DM) continue t...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>22765222</td>\n",
       "      <td>Diversity of 5S rRNA genes within individual p...</td>\n",
       "      <td>We examined intragenomic variation of paralogo...</td>\n",
       "      <td>chemistry classification genetics chemistry ge...</td>\n",
       "      <td>Bacteria DNA, Ribosomal Databases, Nucleic Aci...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>24401686</td>\n",
       "      <td>Myoinositol and glutamate complex neurometabol...</td>\n",
       "      <td>To obtain quantitative neurometabolite measure...</td>\n",
       "      <td>analogs &amp; derivatives metabolism metabolism me...</td>\n",
       "      <td>Adult Aspartic Acid Brain Injuries Case-Contro...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>22914093</td>\n",
       "      <td>Antibiotics in early life alter the murine col...</td>\n",
       "      <td>Antibiotics administered in low doses have bee...</td>\n",
       "      <td>drug effects physiology administration &amp; dosag...</td>\n",
       "      <td>Adiposity Age Factors Animals Anti-Bacterial A...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>23426830</td>\n",
       "      <td>Elevated serum anti-Müllerian hormone in adole...</td>\n",
       "      <td>Serum anti-Müllerian hormone (AMH) is linked t...</td>\n",
       "      <td>blood pathology ultrasonography blood ultrason...</td>\n",
       "      <td>Adolescent Anti-Mullerian Hormone Child Female...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>24344399</td>\n",
       "      <td>Association of obesity-mediated insulin resist...</td>\n",
       "      <td>The hypothalamus is important in hunger and me...</td>\n",
       "      <td>blood analysis anatomy &amp; histology blood blood...</td>\n",
       "      <td>Adolescent Adult Brain-Derived Neurotrophic Fa...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pmid                                              title  \\\n",
       "0  23195993  Gene expression profiles in peripheral blood m...   \n",
       "1  25077433  SOX2 is a cancer-specific regulator of tumour ...   \n",
       "2  24107601  Imaging and cerebrospinal fluid biomarkers in ...   \n",
       "3  24891029  Preliminary evidence of cognitive and brain ab...   \n",
       "4  21691448  Obese Adolescents with Type 2 Diabetes Mellitu...   \n",
       "5  22765222  Diversity of 5S rRNA genes within individual p...   \n",
       "6  24401686  Myoinositol and glutamate complex neurometabol...   \n",
       "7  22914093  Antibiotics in early life alter the murine col...   \n",
       "8  23426830  Elevated serum anti-Müllerian hormone in adole...   \n",
       "9  24344399  Association of obesity-mediated insulin resist...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Occupational exposure to nickel (Ni) is associ...   \n",
       "1  Although the principles that balance stem cell...   \n",
       "2  The pathophysiological process of Alzheimer's ...   \n",
       "3  To ascertain whether pediatric obesity without...   \n",
       "4  The rates of type 2 diabetes (T2DM) continue t...   \n",
       "5  We examined intragenomic variation of paralogo...   \n",
       "6  To obtain quantitative neurometabolite measure...   \n",
       "7  Antibiotics administered in low doses have bee...   \n",
       "8  Serum anti-Müllerian hormone (AMH) is linked t...   \n",
       "9  The hypothalamus is important in hunger and me...   \n",
       "\n",
       "                                     qualifier_terms  \\\n",
       "0  genetics metabolism epidemiology drug effects ...   \n",
       "1  genetics metabolism pathology metabolism patho...   \n",
       "2  cerebrospinal fluid genetics radionuclide imag...   \n",
       "3          pathology physiopathology physiopathology   \n",
       "4                                                      \n",
       "5  chemistry classification genetics chemistry ge...   \n",
       "6  analogs & derivatives metabolism metabolism me...   \n",
       "7  drug effects physiology administration & dosag...   \n",
       "8  blood pathology ultrasonography blood ultrason...   \n",
       "9  blood analysis anatomy & histology blood blood...   \n",
       "\n",
       "                                    descriptor_terms label  \n",
       "0  Adult Asian Continental Ancestry Group Biomark...    T0  \n",
       "1  Animals Carcinoma, Squamous Cell Cell Line, Tu...    T0  \n",
       "2  Aged Aged, 80 and over Alzheimer Disease Amylo...    T0  \n",
       "3  Adolescent Attention Brain Cognition Executive...    T0  \n",
       "4                                                       T0  \n",
       "5  Bacteria DNA, Ribosomal Databases, Nucleic Aci...    T0  \n",
       "6  Adult Aspartic Acid Brain Injuries Case-Contro...    T0  \n",
       "7  Adiposity Age Factors Animals Anti-Bacterial A...    T0  \n",
       "8  Adolescent Anti-Mullerian Hormone Child Female...    T0  \n",
       "9  Adolescent Adult Brain-Derived Neurotrophic Fa...    T0  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check that column names are the same\n",
    "print df_val.shape\n",
    "print df_val.columns.values\n",
    "print df_val['label'].unique()\n",
    "df_val[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Optional: Remove 'TX' scores to test performance\n",
    "df_val = df_val[df_val['label'] != 'TX']\n",
    "print len(df_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cleaning and parsing the test set abstracts...\n",
      "\n",
      "Adding tfidf weights and converting to Bag of Words...\n"
     ]
    }
   ],
   "source": [
    "#Moving on to test set...\n",
    "# Create an empty list and append the clean reviews one by one\n",
    "clean_test_reviews_val = [] \n",
    "\n",
    "print \"Cleaning and parsing the test set abstracts...\\n\"\n",
    "for i in range(df_val[\"abstract\"].size):    \n",
    "    words = KaggleWord2VecUtility.review_to_words( df_val[\"abstract\"].iloc[i], remove_stopwords=stops)\n",
    "    words = words + ' ' + KaggleWord2VecUtility.review_to_words( df_val[\"title\"].iloc[i], remove_stopwords=stops)\n",
    "    if mesh_terms:\n",
    "        words = words + ' ' + KaggleWord2VecUtility.review_to_words( df_val[\"descriptor_terms\"].iloc[i], remove_stopwords=stops)\n",
    "        words = words + ' ' + KaggleWord2VecUtility.review_to_words( df_val[\"qualifier_terms\"].iloc[i], remove_stopwords=stops)\n",
    "    clean_test_reviews_val.append(words)\n",
    "\n",
    "print \"Adding tfidf weights and converting to Bag of Words...\"\n",
    "# Get a bag of words for the test set, and convert to a numpy array\n",
    "test_data_features_val = vectorizer.transform(clean_test_reviews_val)\n",
    "test_data_features_val = test_data_features_val.toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using the random forest to make sentiment label predictions...\n",
      "Scoring the test set\n",
      "{0: 0.59008487654320996, 1: 0.74125874125874125, 2: 0.67950081833060549}\n"
     ]
    }
   ],
   "source": [
    "print \"Using the random forest to make sentiment label predictions...\"\n",
    "# Use the random forest to make sentiment label predictions\n",
    "#using forest fitted on training data\n",
    "result_val = forest.predict(test_data_features_val)\n",
    "\n",
    "x_test_val = label_binarize(df_val['label'], classes=['T0', 'T1/T2', 'T3/T4'])\n",
    "\n",
    "print \"Scoring the test set\"\n",
    "# Compute ROC curve and ROC area for each class\n",
    "fpr_val = dict()\n",
    "tpr_val = dict()\n",
    "roc_auc_val = dict()\n",
    "for i in range(n_classes):\n",
    "    fpr_val[i], tpr_val[i], _ = roc_curve(x_test_val[:, i], result_val[:, i])\n",
    "    roc_auc_val[i] = auc(fpr_val[i], tpr_val[i])\n",
    "\n",
    "print roc_auc_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#our vocab is in the vectorizer- same for all\n",
    "pickle.dump(vectorizer, open('/mnt/mypartition/pubmed_nlp_research/C_article_replication_and_translational_classification/try2-noMesh/vectorizer.p', 'wb'))\n",
    "pickle.dump(forest, open('/mnt/mypartition/pubmed_nlp_research/C_article_replication_and_translational_classification/try2-noMesh/forest.p', 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now create VIP graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5000\n",
      "5000\n",
      "45\n",
      "45\n",
      "45\n",
      "5000\n",
      "5000\n",
      "26\n",
      "26\n",
      "26\n",
      "5000\n",
      "5000\n",
      "41\n",
      "41\n",
      "41\n"
     ]
    }
   ],
   "source": [
    "getplots(0, 'T0')\n",
    "getplots(1, 'T1T2')\n",
    "getplots(2, 'T3T4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def getplots(pos, lab):\n",
    "    # Take a look at the words in the vocabulary-- repeat for each position in feats\n",
    "    vocab = vectorizer.get_feature_names()\n",
    "    print len(vocab)\n",
    "    print len(feats[pos])\n",
    "    inxs = np.where(feats[pos] > 0.004)\n",
    "    print len(inxs[0])\n",
    "    features = feats[pos][inxs]\n",
    "    print len(features)\n",
    "    voc = [vocab[i] for (x,i) in np.ndenumerate(inxs)]\n",
    "    print len(voc)\n",
    "\n",
    "    plt.figure(figsize=(6 * 1.618, 6))\n",
    "    index = np.arange(len(voc))\n",
    "    bar_width = 0.35\n",
    "    plt.bar(index, features, color = 'black', alpha=0.5)\n",
    "    plt.xlabel('features')\n",
    "    plt.ylabel('importance')\n",
    "    plt.title('Feature importances ' + lab)\n",
    "    plt.xticks(index + bar_width, voc)\n",
    "    locs, labels = plt.xticks()\n",
    "    plt.setp(labels, rotation = 90)\n",
    "    plt.tight_layout()\n",
    "    #plt.show()\n",
    "    plt.savefig('/mnt/mypartition/pubmed_nlp_research/C_article_replication_and_translational_classification/try2-noMesh/vips' + lab + '.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import re\n",
    "import nltk\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from bs4 import BeautifulSoup\n",
    "from nltk.corpus import stopwords\n",
    "\n",
    "\n",
    "class KaggleWord2VecUtility(object):\n",
    "    \"\"\"KaggleWord2VecUtility is a utility class for processing raw HTML text into segments for further learning\"\"\"\n",
    "\n",
    "    @staticmethod\n",
    "    def review_to_wordlist( review, remove_stopwords=False ):\n",
    "        # Function to convert a document to a sequence of words,\n",
    "        # optionally removing stop words.  Returns a list of words.\n",
    "        #\n",
    "        # 1. Remove HTML\n",
    "        review_text = BeautifulSoup(review).get_text()\n",
    "        #\n",
    "        # 2. Remove non-letters\n",
    "        review_text = re.sub(\"[^a-zA-Z]\",\" \", review_text)\n",
    "        #\n",
    "        # 3. Convert words to lower case and split them\n",
    "        words = review_text.lower().split()\n",
    "        #\n",
    "        # 4. Optionally remove stop words (false by default)\n",
    "        if remove_stopwords:\n",
    "            stops = set(stopwords.words(\"english\"))\n",
    "            words = [w for w in words if not w in stops]\n",
    "        #\n",
    "        # 5. Return a list of words\n",
    "        return(words)\n",
    "    \n",
    "    @staticmethod\n",
    "    def review_to_words( review, remove_stopwords=False ):\n",
    "        # Function to convert a raw review to a string of words\n",
    "        # The input is a single string (a raw movie review), and \n",
    "        # the output is a single string (a preprocessed movie review)\n",
    "        #\n",
    "        # 1. Remove HTML\n",
    "        review_text = BeautifulSoup(review).get_text() \n",
    "        #\n",
    "        # 2. Remove non-letters        \n",
    "        review_text = re.sub(\"[^a-zA-Z]\", \" \", review_text) \n",
    "        #\n",
    "        # 3. Convert to lower case, split into individual words\n",
    "        words = review_text.lower().split()                             \n",
    "        #\n",
    "        # 4. Optionally remove stop words (false by default)\n",
    "        if remove_stopwords:\n",
    "            stops = set(stopwords.words(\"english\"))\n",
    "            words = [w for w in words if not w in stops]   \n",
    "        #\n",
    "        # 6. Join the words back into one string separated by space, \n",
    "        # and return the result.\n",
    "        return( \" \".join( words ))   \n",
    "\n",
    "    # Define a function to split a review into parsed sentences\n",
    "    @staticmethod\n",
    "    def review_to_sentences( review, tokenizer, remove_stopwords=False ):\n",
    "        # Function to split a review into parsed sentences. Returns a\n",
    "        # list of sentences, where each sentence is a list of words\n",
    "        #\n",
    "        # 1. Use the NLTK tokenizer to split the paragraph into sentences\n",
    "        raw_sentences = tokenizer.tokenize(review.decode('utf8').strip())\n",
    "        #\n",
    "        # 2. Loop over each sentence\n",
    "        sentences = []\n",
    "        for raw_sentence in raw_sentences:\n",
    "            # If a sentence is empty, skip it\n",
    "            if len(raw_sentence) > 0:\n",
    "                # Otherwise, call review_to_wordlist to get a list of words\n",
    "                sentences.append( KaggleWord2VecUtility.review_to_wordlist( raw_sentence, \\\n",
    "                  remove_stopwords ))\n",
    "        #\n",
    "        # Return the list of sentences (each sentence is a list of words,\n",
    "        # so this returns a list of lists\n",
    "        return sentences"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
